### PR TITLE
feat: Show outdated Thunderstore mods first

### LINKED ISSUE #293: Sort Thunderstore mod browser by outdated first
Right now there's not an easy way in FlightCore to see which mods are outdated. The optimal way would be to update the local mods view to show outdated Thunderstore mods.

Until this is done (I'm rather busy atm) we could also update the sort function in the Thunderstore mod browser view to sort by outdated first and then by upload date.

This way opening Thunderstore mod browser would instantly show all outdated mods ^^

### PR DESCRIPTION
**Pull request by Alystrasz**
* A badge displays count of outdated mods
* Outdated mods are always displayed first

image

Closes #293.

### CODE REVIEW

File: src-vue/src/components/ModsMenu.vue [hunk 1]
```diff
@@ -50,6 +58,8 @@
 <script lang="ts">
 import { defineComponent } from 'vue'
 import { SortOptions } from '../utils/SortOptions.d';
+import {isThunderstoreModOutdated} from "../utils/thunderstore/version";
+import {ThunderstoreMod} from "../../../src-tauri/bindings/ThunderstoreMod";
```
Comment (GeckoEidechse) [inline src-vue/src/components/ModsMenu.vue:62]: Formatting :P

```suggestion
import { isThunderstoreModOutdated } from "../utils/thunderstore/version";
import { ThunderstoreMod } from "../../../src-tauri/bindings/ThunderstoreMod";
```

File: src-vue/src/utils/thunderstore/version.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import {ThunderstoreMod} from "../../../../src-tauri/bindings/ThunderstoreMod";
+import {NorthstarMod} from "../../../../src-tauri/bindings/NorthstarMod";
+import {store} from "../../plugins/store";
```
Comment (GeckoEidechse) [inline src-vue/src/utils/thunderstore/version.ts:3]:
```suggestion
import { ThunderstoreMod } from "../../../../src-tauri/bindings/ThunderstoreMod";
import { NorthstarMod } from "../../../../src-tauri/bindings/NorthstarMod";
import { store } from "../../plugins/store";
```

File: src-vue/src/utils/thunderstore/version.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * Strips off a Thunderstore dependency string from its version
+ * (e.g. "taskinoz-WallrunningTitans-1.0.0" to
+ * "taskinoz-WallrunningTitans").
+ **/
+function getThunderstoreDependencyStringPrefix(dependency: string): string {
```
Comment (GeckoEidechse) [inline src-vue/src/utils/thunderstore/version.ts:10]: I'll probably move this to backend functions at some point but for the sake of getting this out ASAP I'm fine with having it in frontend for now ^^

File: src-vue/src/views/mods/ThunderstoreModsView.vue [hunk 4]
```diff
@@ -52,6 +52,7 @@ import { ElScrollbar, ScrollbarInstance } from "element-plus";
 import { SortOptions } from "../../utils/SortOptions.d";
 import { ThunderstoreModVersion } from "../../../../src-tauri/bindings/ThunderstoreModVersion";
 import { fuzzy_filter } from "../../utils/filter";
+import {isThunderstoreModOutdated} from "../../utils/thunderstore/version";
```
Comment (GeckoEidechse) [inline src-vue/src/views/mods/ThunderstoreModsView.vue:55]:
```suggestion
import { isThunderstoreModOutdated } from "../../utils/thunderstore/version";

```

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]
Yoo, this is really awesome :D

First review, formatting stuff for now ^^

File: src-vue/src/utils/thunderstore/version.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    const dependencyStringMembers = dependency.split('-');
+    return `${dependencyStringMembers[0]}-${dependencyStringMembers[1]}`;
+}
+
+function isThunderstoreModOutdated(mod: ThunderstoreMod): boolean {
+    // Ensure mod is up-to-date.
+    const tsModPrefix = getThunderstoreDependencyStringPrefix(mod.versions[0].full_name);
+    const matchingMods: NorthstarMod[] = store.state.installed_mods.filter((mod: NorthstarMod) => {
+        if (!mod.thunderstore_mod_string) return false;
+        return getThunderstoreDependencyStringPrefix(mod.thunderstore_mod_string!) === tsModPrefix;
```
Comment (GeckoEidechse) [inline src-vue/src/utils/thunderstore/version.ts:20]: What does the trailing exclamation mark do?

File: src-vue/src/utils/thunderstore/version.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    });
+    if (matchingMods.length !== 0) {
+        // There shouldn't be several mods with same dependency string, but we never know...
+        const matchingMod = matchingMods[0];
+        // A mod is outdated if its dependency strings differs from Thunderstore dependency string
+        // (no need for semver check here)
+        return matchingMod.thunderstore_mod_string !== mod.versions[0].full_name;
```
Comment (GeckoEidechse) [inline src-vue/src/utils/thunderstore/version.ts:27]: Maybe make a comment that this assumes mod versions list is always sorted newest->oldest which I'm not sure is guaranteed... ^^"

File: src-vue/src/views/mods/ThunderstoreModsView.vue [hunk 7]
```diff
@@ -137,7 +138,17 @@ export default defineComponent({
                     throw new Error('Unknown mod sorting.');
             }
 
-            return mods.sort(compare);
+            // Display outdated mods first
```
Comment (GeckoEidechse) [inline src-vue/src/views/mods/ThunderstoreModsView.vue:141]:
```suggestion
            // Always display outdated mods first
            // (regardless of actual sort order)
```

**Review by GeckoEidechse — APPROVED** [review]
Works in testing :D

So as soon as comments are resolved we should be ready to merge :tada:

File: src-vue/src/utils/thunderstore/version.ts (hunk 6 above)
Comment (Alystrasz) [reply src-vue/src/utils/thunderstore/version.ts:28]: Addressed in e61463e14deb85bda972ede5b2c0a8b715e9c855

File: src-vue/src/utils/thunderstore/version.ts (hunk 5 above)
Comment (Alystrasz) [reply src-vue/src/utils/thunderstore/version.ts:20]: It ensures that `thunderstore_mod_string` is not null

Comment (GeckoEidechse) [reply src-vue/src/utils/thunderstore/version.ts:20]: Ah, it's a TypeScript thing to tell the compiler that value will never be `null` here, right?

Comment (Alystrasz) [reply src-vue/src/utils/thunderstore/version.ts:20]: yep!

**Pull request merged**